sdc/extensions/sdc_hashmap_ext.py

```diff
@@ -0,0 +1,1123 @@
+    key_type_postfix, value_type_postfix = _get_types_postfixes(ty_key, ty_val)
+
+    def codegen(context, builder, sig, args):
+        dict_val, key_val = args
+
+        key_val, lir_key_type = transform_input_arg(context, builder, key_type, key_val)
+        native_value_ptr, lir_value_type = alloc_native_value(context, builder, ty_val)
+
+        cdict = cgutils.create_struct_proxy(dict_type)(context, builder, value=dict_val)
+        fnty = lir.FunctionType(lir.IntType(8),
+                                [lir.IntType(8).as_pointer(),
+                                 lir_key_type,
+                                 lir_value_type.as_pointer()
+                                 ])
+        func_name = f"hashmap_lookup_{key_type_postfix}_to_{value_type_postfix}"
+        fn_hashmap_lookup = builder.module.get_or_insert_function(
+            fnty, name=func_name)
+
+        status = builder.call(fn_hashmap_lookup, [cdict.data_ptr, key_val, native_value_ptr])
+        status_as_bool = context.cast(builder, status, types.uint8, types.bool_)
+
+        # if key was not found nothing would be stored to native_value_ptr, so depending on status
+        # we either deref it or not, wrapping final result into types.Optional value
+        result_ptr = cgutils.alloca_once(builder,
+                                         context.get_value_type(types.Optional(ty_val)))
+        with builder.if_else(status_as_bool, likely=True) as (if_ok, if_not_ok):
+            with if_ok:
+                native_value = builder.load(native_value_ptr)
+                result_value = transform_native_val(context, builder, ty_val, native_value)
+
+                if context.enable_nrt:
+                    context.nrt.incref(builder, ty_val, result_value)
+
+                builder.store(context.make_optional_value(builder, ty_val, result_value),
+                              result_ptr)
+
+            with if_not_ok:
+                builder.store(context.make_optional_none(builder, ty_val),
+                              result_ptr)
+
+        opt_result = builder.load(result_ptr)
+        return context.make_tuple(builder, return_type, [status_as_bool, opt_result])
+
+    func_sig = return_type(dict_type, key_type)
+    return func_sig, codegen
+
+
+@overload(operator.getitem, prefer_literal=False)
+def concurrent_dict_lookup_ovld(self, key):
+    if not isinstance(self, ConcurrentDictType):
+        return None
+
+    dict_key_type = self.key_type
+    cast_key = key is not dict_key_type
+
+    def concurrent_dict_lookup_impl(self, key):
+        _key = key if cast_key == False else _cast(key, dict_key_type)  # noqa
+        found, res = hashmap_lookup(self, _key)
+
+        # Note: this function raises exception so expect no scaling if you use it in prange
+        if not found:
+            raise KeyError("ConcurrentDict key not found")
+        return res
+
+    return concurrent_dict_lookup_impl
+
+
+@intrinsic
+def hashmap_clear(typingctx, dict_type):

```

Comment:
why do we need this to be as intrinsic and not as an overload?

```python
@overload(hashmap_clear)
def hashmap_clear_ovld(conc_dict):
    ty_key, ty_val = dict_type.key_type, dict_type.value_type
    key_type_postfix, value_type_postfix = _get_types_postfixes(ty_key, ty_val)

    hasmap_clear_func = hashmap_clear_map[(key_type_postfix, value_type_postfix)]
    
    def hashmap_clear_impl(conc_dict):
        return hashmap_clear_func(conc_dict._data_ptr)

    return hashmap_clear_impl
```

Am I missing something?

ctypes binding is required for this

Reply:
Thanks! I didn't consider this way of doing things. However it may be more difficult to read for methods like `set` and `pop` which make key transformations during codegen or allocate the result (e.g. https://github.com/IntelPython/sdc/blob/11774955fb3ce3237d89bc59c3522b99d6c61e55/sdc/extensions/sdc_hashmap_ext.py#L434), since the logic of generated llvm code will be split across functions (and also it would be much harder to debug this if you need to make some changed in the codegen). So if you don't mind I will keep it initial way for now.